daml-assistant/daml-helper/test/DA/Daml/Helper/Test/Deployment.hs

```diff
@@ -0,0 +1,128 @@
+  -- so running tests in parallel will cause trouble.
+  setEnv "TASTY_NUM_THREADS" "1" True
+  helper <- locateRunfiles (mainWorkspace </> "daml-assistant" </> "daml-helper")
+  damlc <- locateRunfiles (mainWorkspace </> "compiler" </> "damlc")
+  let path = intercalate [searchPathSeparator] ["/bin", "/usr/bin", damlc, helper]
+  withEnv [ ("PATH", Just path)] $ defaultMain $ testGroup "DAML-UX"
+    [ authenticatedUploadTest
+    , fetchTest
+    ]
+
+-- | Test `daml ledger upload-dar --access-token-file`
+authenticatedUploadTest :: TestTree
+authenticatedUploadTest = do
+  let sharedSecret = "TheSharedSecret"
+  withSandbox defaultSandboxConf { mbSharedSecret = Just sharedSecret } $ \getSandboxPort ->
+    testCaseSteps "authenticatedUploadTest" $ \step -> do
+    port <- getSandboxPort
+    withTempDir $ \deployDir -> do
+      withCurrentDirectory deployDir $ do
+        writeMinimalProject
+        step "build"
+        callCommandQuiet "damlc build"
+        let dar = ".daml/dist/proj1-0.0.1.dar"
+        let tokenFile = deployDir </> "secretToken.jwt"
+        step "upload"
+        -- The trailing newline is not required but we want to test that it is supported.
+        writeFileUTF8 tokenFile ("Bearer " <> makeSignedJwt sharedSecret <> "\n")
+        callCommandQuiet $ unwords
+          [ "daml-helper ledger upload-dar"
+          , "--access-token-file " <> tokenFile
+          , "--host localhost" , "--port", show port
+          , dar
+          ]
+
+makeSignedJwt :: String -> String
+makeSignedJwt sharedSecret = do
+  let urc = JWT.ClaimsMap $ Map.fromList [ ("admin", Aeson.Bool True)]
+  let cs = mempty { JWT.unregisteredClaims = urc }
+  let key = JWT.hmacSecret $ T.pack sharedSecret
+  let text = JWT.encodeSigned key mempty cs
+  T.unpack text
+
+-- | Test `daml ledger fetch-dar`
+fetchTest :: TestTree
+fetchTest = do
+  withSandbox defaultSandboxConf $ \getSandboxPort ->
+    testCaseSteps "fetchTest" $ \step -> do
+    port <- getSandboxPort
+    withTempDir $ \fetchDir -> do
+      withCurrentDirectory fetchDir $ do
+        writeMinimalProject
+        let origDar = ".daml/dist/proj1-0.0.1.dar"
+        step "build/upload"
+        callCommandQuiet $ unwords ["damlc build"]
+        callCommandQuiet $ unwords
+          [ "daml-helper ledger upload-dar"
+          , "--host", "localhost" , "--port" , show port
+          , origDar
+          ]
+        pid <- getMainPidByInspecingDar origDar "proj1"
+        step "fetch/validate"
+        let fetchedDar = "fetched.dar"
+        callCommandQuiet $ unwords
+          [ "daml-helper ledger fetch-dar"
+          , "--host", "localhost" , "--port", show port
+          , "--main-package-id", pid
+          , "-o", fetchedDar
+          ]
+        callCommandQuiet $ unwords ["damlc validate-dar", fetchedDar]
+
+-- | Using `daml inspect-dar`, discover the main package-identifier of a dar.
+getMainPidByInspecingDar :: FilePath -> String -> IO String
```

Comment:
As I said on the other PR, this is just unnecessarily hacky. Let’s please just read the DAR like we do in the packaging tests instead of calling out to `damlc inspect-dar`.

Reply:
I don't think the `inspect-dar` is that hacky! It was the simplest user level way I could think of to determine the `pkgId` from a DAR.

However, I think what it highlights is that since we now have a user facing command which takes a `pkgId` argument, we also need a user facing command which returns the package-id of a DAR. This would be used here instead of the `inspect-dar` hack.

In any case, regardless of the hackiness of using `inspect-dar`, it pre-existed this PR, and is not the focus of this PR which is to move testing out of integration-tests.

Reply:
You don’t need a user-level way, using a library function is perfectly fine here. And claiming that it predates this PR is a bit silly. I pointed this out in the PR where you added it and agreed to refactor the tests separately under the condition that the comments are addressed.

I do agree that we need a machine readable output for `inspect-dar`, see https://github.com/digital-asset/daml/issues/3117 but even if we had that, I don’t want to use this here. Tying five different commands together only makes it harder to figure out what is actually going wrong if the tests fails. Tests should focus on testing one thing not test the `inspect-dar` command when they really want to test `fetch-dar`.

Please address this as part of this PR.